--- a/openstack_dashboard/dashboards/project/access_and_security/api_access/views.py	2013-06-19 14:40:57.000000000 +0200
+++ b/openstack_dashboard/dashboards/project/access_and_security/api_access/views.py	2013-06-19 14:41:22.000000000 +0200
@@ -97,7 +97,7 @@
     response = http.HttpResponse(mimetype='application/zip')
     response.write(temp_zip.read())
     response['Content-Disposition'] = ('attachment; '
-                                       'filename=%s-x509.zip'
+                                       'filename="%s-x509.zip"'
                                        % tenant_name)
     response['Content-Length'] = temp_zip.tell()
     return response
@@ -124,7 +124,7 @@
                                     context,
                                     content_type="text/plain")
         response['Content-Disposition'] = ('attachment; '
-                                           'filename=%s-openrc.sh'
+                                           'filename="%s-openrc.sh"'
                                            % tenant_name)
         response['Content-Length'] = str(len(response.content))
         return response
