From: Jose Castro Leon <jose.castro.leon@cern.ch>
Date: Fri, 21 Feb 2014 08:11:00 +0200
Subject: [PATCH] CERN Disable floating ips and security groups
---

--- a/openstack_dashboard/dashboards/project/instances/tables.py	2013-06-19 14:45:10.000000000 +0200
+++ b/openstack_dashboard/dashboards/project/instances/tables.py	2013-06-19 14:45:31.000000000 +0200
@@ -580,9 +580,7 @@
         table_actions = (LaunchLink, SoftRebootInstance, TerminateInstance,
                          InstancesFilterAction)
         row_actions = (StartInstance, ConfirmResize, RevertResize,
-                       CreateSnapshot, SimpleAssociateIP, AssociateIP,
-                       SimpleDisassociateIP, EditInstance,
-                       EditInstanceSecurityGroups, ConsoleLink, LogLink,
+                       CreateSnapshot, ConsoleLink, LogLink,
                        TogglePause, ToggleSuspend, ResizeLink,
                        SoftRebootInstance, RebootInstance, StopInstance,
                        RebuildInstance, TerminateInstance)
--- a/openstack_dashboard/dashboards/project/access_and_security/tabs.py	2013-06-19 14:47:31.000000000 +0200
+++ b/openstack_dashboard/dashboards/project/access_and_security/tabs.py	2013-06-19 14:47:44.000000000 +0200
@@ -128,5 +128,5 @@

 class AccessAndSecurityTabs(tabs.TabGroup):
     slug = "access_security_tabs"
-    tabs = (SecurityGroupsTab, KeypairsTab, FloatingIPsTab, APIAccessTab)
+    tabs = (KeypairsTab, APIAccessTab)
     sticky = True
