--- a/openstack_dashboard/dashboards/project/access_and_security/templates/access_and_security/api_access/openrc.sh.template	2013-06-19 14:35:46.000000000 +0200
+++ b/openstack_dashboard/dashboards/project/access_and_security/templates/access_and_security/api_access/openrc.sh.template	2013-06-19 14:36:48.000000000 +0200
@@ -14,6 +14,9 @@
 export OS_TENANT_ID={{ tenant_id }}
 export OS_TENANT_NAME="{{ tenant_name }}"

+# CERN SSL Validation CA Chain
+export OS_CACERT=/etc/pki/tls/certs/CERN-bundle.pem
+
 # In addition to the owning entity (tenant), openstack stores the entity
 # performing the action as the **user**.
 export OS_USERNAME={{ user.username }}
