--- a/openstack_dashboard/dashboards/project/instances/tables.py	2013-10-14 11:40:08.000000000 +0200
+++ b/openstack_dashboard/dashboards/project/instances/tables.py	2013-10-14 11:40:24.000000000 +0200
@@ -482,7 +482,6 @@
         row_class = UpdateRow
         table_actions = (LaunchLink, TerminateInstance)
         row_actions = (ConfirmResize, RevertResize, CreateSnapshot,
-                       EditInstance,
-                       EditInstanceSecurityGroups, ConsoleLink, LogLink,
+                       ConsoleLink, LogLink,
                        TogglePause, ToggleSuspend, SoftRebootInstance,
                        RebootInstance, TerminateInstance)
