--- a/openstack_dashboard/dashboards/settings/dashboard.py	2014-02-03 09:26:03.000000001 +0100
+++ b/openstack_dashboard/dashboards/settings/dashboard.py	2014-02-03 09:26:16.000000001 +0100
@@ -23,7 +23,7 @@
 class Settings(horizon.Dashboard):
     name = _("Settings")
     slug = "settings"
-    panels = ('user', 'password', )
+    panels = ('user', )
     default_panel = 'user'
     nav = False
 
