From:  Matthias Runge <mrunge@redhat.com>
Date: Mon, 25 Feb 2014 12:25:13 +0200
Subject: [PATCH] Reduce number of novaclient calls
---
diff --git a/openstack_dashboard/api/nova.py b/openstack_dashboard/api/nova.py
index e30e56f..8d9845d 100644
--- a/openstack_dashboard/api/nova.py
+++ b/openstack_dashboard/api/nova.py
@@ -493,6 +493,7 @@ def server_list(request, search_opts=None, all_tenants=False):
     page_size = request.session.get('horizon_pagesize',
                                     getattr(settings, 'API_RESULT_PAGE_SIZE',
                                             20))
+    c = novaclient(request)
     paginate = False
     if search_opts is None:
         search_opts = {}
@@ -506,7 +507,7 @@ def server_list(request, search_opts=None, all_tenants=False):
     else:
         search_opts['project_id'] = request.user.tenant_id
     servers = [Server(s, request)
-                for s in novaclient(request).servers.list(True, search_opts)]
+                for s in c.servers.list(True, search_opts)]
 
     has_more_data = False
     if paginate and len(servers) > page_size:
@@ -670,8 +671,9 @@ def service_list(request):
 
 def aggregate_list(request):
     result = []
-    for aggregate in novaclient(request).aggregates.list():
-        result.append(novaclient(request).aggregates.get_details(aggregate.id))
+    c = novaclient(request)
+    for aggregate in c.aggregates.list():
+        result.append(c.aggregates.get_details(aggregate.id))
 
     return result
 
