From: Jose Castro Leon <jose.castro.leon@cern.ch>
Date: Thu, 22 May 2014 13:02:00 +0200
Subject: [PATCH] CERN Enable RDP Console backport from blueprint
---
--- a/openstack_dashboard/api/nova.py
+++ b/openstack_dashboard/api/nova.py
@@ -65,6 +65,13 @@ class SPICEConsole(base.APIDictWrapper):
     _attrs = ['url', 'type']
 
 
+class RDPConsole(base.APIDictWrapper):
+    """Wrapper for the "console" dictionary returned by the
+    novaclient.servers.get_rdp_console method.
+    """
+    _attrs = ['url', 'type']
+
+
 class Server(base.APIResourceWrapper):
     """Simple wrapper around novaclient.server.Server
 
@@ -385,6 +392,11 @@ def server_spice_console(request, instance_id, console_type='spice-html5'):
         instance_id, console_type)['console'])
 
 
+def server_rdp_console(request, instance_id, console_type='rdp-html5'):
+    return RDPConsole(novaclient(request).servers.get_rdp_console(
+        instance_id, console_type)['console'])
+
+
 def flavor_create(request, name, memory, vcpu, disk, flavorid='auto',
                   ephemeral=0, swap=0, metadata=None, is_public=True):
     flavor = novaclient(request).flavors.create(name, memory, vcpu, disk,
--- a/openstack_dashboard/dashboards/admin/instances/urls.py
+++ b/openstack_dashboard/dashboards/admin/instances/urls.py
@@ -37,4 +37,5 @@
     url(INSTANCES % 'console', 'console', name='console'),
     url(INSTANCES % 'vnc', 'vnc', name='vnc'),
     url(INSTANCES % 'spice', 'spice', name='spice'),
+    url(INSTANCES % 'rdp', 'rdp', name='rdp'),
 )
--- a/openstack_dashboard/dashboards/admin/instances/views.py
+++ b/openstack_dashboard/dashboards/admin/instances/views.py
@@ -54,6 +54,11 @@ def spice(args, **kvargs):
     return views.spice(args, **kvargs)
 
 
+# re-use rdp from project.instances.views to make reflection work
+def rdp(args, **kvargs):
+    return views.rdp(args, **kvargs)
+
+
 class AdminUpdateView(views.UpdateView):
     workflow_class = update_instance.AdminUpdateInstance
 
--- a/openstack_dashboard/dashboards/project/instances/tabs.py
+++ b/openstack_dashboard/dashboards/project/instances/tabs.py
@@ -79,7 +79,15 @@ def get_context_data(self, request):
                         getattr(instance, "name", ""),
                         instance.id)
                 except Exception:
-                    console_url = None
+                    try:
+                        console = api.nova.server_rdp_console(request,
+                                                              instance.id)
+                        console_url = "%s&title=%s(%s)" % (
+                            console.url,
+                            getattr(instance, "name", ""),
+                            instance.id)
+                    except Exception:
+                        console_url = None
         elif console_type == 'VNC':
             try:
                 console = api.nova.server_vnc_console(request, instance.id)
@@ -98,6 +106,15 @@ def get_context_data(self, request):
                     instance.id)
             except Exception:
                 console_url = None
+        elif console_type == 'RDP':
+            try:
+                console = api.nova.server_rdp_console(request, instance.id)
+                console_url = "%s&title=%s(%s)" % (
+                    console.url,
+                    getattr(instance, "name", ""),
+                    instance.id)
+            except Exception:
+                console_url = None
         else:
             console_url = None
 
--- a/openstack_dashboard/dashboards/project/instances/urls.py
+++ b/openstack_dashboard/dashboards/project/instances/urls.py
@@ -38,5 +38,6 @@
     url(INSTANCES % 'console', 'console', name='console'),
     url(INSTANCES % 'vnc', 'vnc', name='vnc'),
     url(INSTANCES % 'spice', 'spice', name='spice'),
+    url(INSTANCES % 'rdp', 'rdp', name='rdp'),
     url(INSTANCES % 'resize', views.ResizeView.as_view(), name='resize'),
 )
--- a/openstack_dashboard/dashboards/project/instances/views.py
+++ b/openstack_dashboard/dashboards/project/instances/views.py
@@ -164,6 +164,18 @@ def spice(request, instance_id):
         exceptions.handle(request, msg, redirect=redirect)
 
 
+def rdp(request, instance_id):
+    try:
+        console = api.nova.server_rdp_console(request, instance_id)
+        instance = api.nova.server_get(request, instance_id)
+        return shortcuts.redirect(console.url +
+                ("&title=%s(%s)" % (instance.name, instance_id)))
+    except Exception:
+        redirect = reverse("horizon:project:instances:index")
+        msg = _('Unable to get RDP console for instance "%s".') % instance_id
+        exceptions.handle(request, msg, redirect=redirect)
+
+
 class UpdateView(workflows.WorkflowView):
     workflow_class = project_workflows.UpdateInstance
     success_url = reverse_lazy("horizon:project:instances:index")
--- a/openstack_dashboard/local/local_settings.py.example
+++ b/openstack_dashboard/local/local_settings.py.example
@@ -44,7 +44,7 @@ TEMPLATE_DEBUG = DEBUG
 # OPENSTACK_KEYSTONE_DEFAULT_DOMAIN = 'Default'
 
 # Set Console type:
-# valid options would be "AUTO", "VNC" or "SPICE"
+# valid options would be "AUTO", "VNC", "SPICE" or "RDP"
 # CONSOLE_TYPE = "AUTO"
 
 # Default OpenStack Dashboard configuration.
